op-plasma/damgr.go

```diff
@@ -2,48 +2,386 @@ package plasma
+			if tx.Hash() != log.TxHash {
+				d.log.Error("tx hash mismatch", "block", block.Number, "tx", i, "log", log.Index, "txHash", tx.Hash(), "receiptTxHash", log.TxHash)
+				continue
+			}
+			// Decode the input from resolver tx calldata
+			input, err := DecodeResolvedInput(tx.Data())
+			if err != nil {
+				d.log.Error("failed to decode resolved input", "block", block.Number, "tx", i, "err", err)
+				continue
+			}
+			if err := comm.Verify(input); err != nil {
+				d.log.Error("failed to verify commitment", "block", block.Number, "tx", i, "err", err)
+				continue
+			}
+			d.log.Debug("resolved input", "block", block.Number, "tx", i)
+			d.state.SetResolvedChallenge(comm.Encode(), input, log.BlockNumber)
+		case ChallengeActive:
+			d.state.SetActiveChallenge(comm.Encode(), log.BlockNumber, d.cfg.ResolveWindow)
+		default:
+			d.log.Warn("skipping unknown challenge status", "block", block.Number, "tx", i, "log", log.Index, "status", status)
+		}
+	}
+	return nil
+}
+
+// fetchChallengeLogs returns logs for challenge events if any for the given block
+func (d *DA) fetchChallengeLogs(ctx context.Context, block eth.BlockID) ([]*types.Log, error) { //cached with deposits events call so not expensive
+	var logs []*types.Log
+	_, receipts, err := d.l1.FetchReceipts(ctx, block.Hash)
+	if err != nil {
+		return logs, err
+	}
+	d.log.Info("loading challenges", "epoch", block.Number, "numReceipts", len(receipts))
+	for _, rec := range receipts {
+		// skip error logs
+		if rec.Status != types.ReceiptStatusSuccessful {
+			continue
+		}
+		for _, log := range rec.Logs {
+			if log.Address == d.cfg.DAChallengeContractAddress && len(log.Topics) > 0 && log.Topics[0] == ChallengeStatusEventABIHash {
+				logs = append(logs, log)
+			}
+		}
+	}
+
+	return logs, nil
+}
+
+// decodeChallengeStatus decodes and validates a challenge event from a transaction log, returning the associated commitment bytes.
+func (d *DA) decodeChallengeStatus(log *types.Log) (ChallengeStatus, Keccak256Commitment, error) {
+	event, err := DecodeChallengeStatusEvent(log)
+	if err != nil {
+		return 0, nil, err
+	}
+	d.log.Debug("decoded challenge status event", "log", log, "event", event)
+	comm, err := DecodeKeccak256(event.ChallengedCommitment)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	bn := event.ChallengedBlockNumber.Uint64()
+	// if we are not tracking the commitment from processing the l1 origin in derivation,
+	// i.e. someone challenged garbage data, this challenge is invalid.
+	if !d.state.IsTracking(comm.Encode(), bn) {
+		return 0, nil, fmt.Errorf("%w: %x at block %d", ErrInvalidChallenge, comm.Encode(), bn)
+	}
+	return ChallengeStatus(event.Status), comm, nil
+}
+
+// LookAhead increments the challenges head and process the new block if it exists.
+// It is only used if the derivation pipeline stalls and we need to wait for a challenge to be resolved
+// to get the next input.
+func (d *DA) LookAhead(ctx context.Context) error {
```

Comment:
It took me a while to figure out how this function was called in a loop to advance to the end of the challenge window.